### PR TITLE
Feature/pipeline caching

### PR DESCRIPTION
**Pull request by mitchdenny**
This PR introduces Pipeline Caching into the Java builds (for CI validation, not PR validation). Eventually we'll do PR validation as well but there is currently a bug preventing us from using it. The PR also adds a control mechanism to turn of Jetty on a per pipeline basis (and also turn off caching as well if it is causing problems).

There is also a CacheSalt variable that can be used to invalidate the cache if something is screwy and we need to rebuild it. This PR will also go some ways to reducing the volume of build failures in Java because we will reduce the number of hits to Maven Central (the packages will come from the cache) although it won't completely eliminate them.

### CODE REVIEW

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 1]
```diff
@@ -78,7 +81,7 @@ jobs:
           javaHomeOption: 'JDKVersion'
           jdkVersionOption: '1.11'
           jdkArchitectureOption: 'x64'
-          publishJUnitResults: false
+          publishJUnitResults: false3
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml]: looks like the extra 3 is a bad edit.

File: eng/pipelines/templates/steps/cache-maven-repository.yml [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    echo "##vso[task.setvariable variable=Maven.RepositoryPath;]$HOME/.m2/repository"
+  condition: and(ne(variables['Agent.OS'], 'Windows_NT'), ne(variables['Build.Reason'], 'PullRequest'), eq(variables['EnableCache'], 'true'))
+  displayName: Detecting Maven repository on Linux and macOS
+- task: CacheBeta@0
+  inputs:
+    key: |
+      JobName-$(Agent.JobName)
+      CacheSalt-$(CacheSalt)
+      EnableJetty-$(EnableJetty)
+      $(Build.SourcesDirectory)/pom.client.xml
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/cache-maven-repository.yml]: ugh... you have to explicitly list all the files?

File: eng/pipelines/templates/steps/cache-maven-repository.yml [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+  displayName: Detecting Maven repository on Windows
+- script: |
+    echo "##vso[task.setvariable variable=Maven.RepositoryPath;]$HOME/.m2/repository"
+  condition: and(ne(variables['Agent.OS'], 'Windows_NT'), ne(variables['Build.Reason'], 'PullRequest'), eq(variables['EnableCache'], 'true'))
+  displayName: Detecting Maven repository on Linux and macOS
+- task: CacheBeta@0
+  inputs:
+    key: |
+      JobName-$(Agent.JobName)
+      CacheSalt-$(CacheSalt)
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/cache-maven-repository.yml]: Where does CacheSalt come from?

File: eng/pipelines/templates/steps/cache-maven-repository.yml (hunk 2 above)
Comment (mitchdenny) [reply eng/pipelines/templates/steps/cache-maven-repository.yml]: Currently - yes, there is a top feature request for pipeline caching to support globbing here. For the time being I thought I would centralize this horror of a fingerprint list in a separate template that we can just reference in ;) Fingerprinting all these files takes something like 2 seconds (tops). Hopefully by the time we take a second pass at this we can just replace it with a glob!

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml (hunk 1 above)
Comment (mitchdenny) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Fixed.

File: eng/pipelines/templates/steps/cache-maven-repository.yml (hunk 3 above)
Comment (mitchdenny) [reply eng/pipelines/templates/steps/cache-maven-repository.yml]: Its a variable on the pipeline. Until they have a cache invalidation switch I've included it here so we can just update a variable on the pipeline definition to invalidate it forcibly.

File: eng/pipelines/templates/steps/cache-maven-repository.yml (hunk 2 above)
Comment (weshaggard) [reply eng/pipelines/templates/steps/cache-maven-repository.yml]: It should be pretty easy to replace it with a glob in a script block and set it as a variable we use in this block.

File: eng/pipelines/templates/steps/cache-maven-repository.yml [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+      EnableJetty-$(EnableJetty)
+      $(Build.SourcesDirectory)/pom.client.xml
+      $(Build.SourcesDirectory)/pom.data.xml
+      $(Build.SourcesDirectory)/pom.management.xml
+      $(Build.SourcesDirectory)/pom.xml
+      $(Build.SourcesDirectory)/advisor/resource-manager/v2017_04_19/pom.xml
```
Comment (willsmythe) [inline eng/pipelines/templates/steps/cache-maven-repository.yml]: @mitchdenny - key segments can now contain wildcards. This should help reduce the number of file paths you need to list here. For example:

`key: JobName-$(Agent.JobName) | CacheSalt-$(CacheSalt) | EnableJetty-$(EnableJetty) | **/pom*.xml`

Comment (mitchdenny) [reply eng/pipelines/templates/steps/cache-maven-repository.yml]: Thanks Will. I have another PR where I have this all setup and it is working. It does improve perf although:

**Pull request closed without merging**